src/main/java/teammates/ui/controller/Action.java

```diff
@@ -238,10 +238,9 @@ protected AccountAttributes authenticateAndGetNominalUser(UserType loggedInUserT
                 
                 throw new UnauthorizedAccessException("Unregistered user for a page that needs one");
             } else if (isPageNotCourseJoinRelated() && doesRegkeyBelongToUnregisteredStudent() && isUserLoggedIn) {
-                Url redirectUrl = new Url(Const.ViewURIs.LOGOUT)
-                                      .withParam(Const.ParamsNames.NEXT_URL, requestUrl)
-                                      .withUserId(StringHelper.encrypt(account.googleId)); 
-                
+                Url redirectUrl = new Url(student.getRegistrationUrl())
+                                      .withParam(Const.ParamsNames.STUDENT_NEW_COURSE, student.course)
+                                      .withParam(Const.ParamsNames.NEXT_URL, requestUrl);
```

Comment:
If you mean the user is a total stranger, then it would be captured by the block just before it (`doesUserNeedRegistration(account)`).
If you mean the user was registered before but his/her courses were all deleted, then yes 3 includes such case as well. But really what are the chances of this happening.